okhttp/src/main/java/okhttp3/internal/http/CallServerInterceptor.java

```diff
@@ -82,26 +107,34 @@ public CallServerInterceptor(boolean forWebSocket) {
         .receivedResponseAtMillis(System.currentTimeMillis())
         .build();
 
+    if ("close".equalsIgnoreCase(response.request().header("Connection"))
+        || "close".equalsIgnoreCase(response.header("Connection"))) {
+      streamAllocation.noNewStreams();
+    }
+
     int code = response.code();
     if (forWebSocket && code == 101) {
       // Connection is upgrading, but we need to ensure interceptors see a non-null response body.
       response = response.newBuilder()
           .body(Util.EMPTY_RESPONSE)
           .build();
     } else {
-      response = response.newBuilder()
-          .body(httpCodec.openResponseBody(response))
-          .build();
-    }
+      realChain.eventListener().responseBodyStart(realChain.call());
+      try {
+        response = response.newBuilder()
+            .body(httpCodec.openResponseBody(response))
```

Comment:
Is it worth rethinking the EventListener and introduce parameter objects to allow additional fields in future without breaking compatibility?

Or marking the EventListener API as clearly experimental and likely to change?


```
  public void connectEnd(Call call, @Nullable ConnectEndDetails details, @Nullable Throwable throwable) {
  }

  public void responseBodyEnd(Call call, @Nullable ResponseBodyDetails details, @Nullable Throwable throwable) {
  }
```